037-kafka-canary-roadmap.md

```diff
@@ -0,0 +1,106 @@
+So generally speaking, we would expect the Canary to experience fewer issues due to client bugs.
+It would also necessarily support the latest versions of the Kafka protocol, which would enable support for more features.
+
+The Canary also stands out as being the only Strimzi project that's written in golang.
+There's nothing wrong with golang, but there are non-technical drawbacks to this situation.
+Most existing Strimzi developers have more experience with Java than golang.
+Some don't really know golang at all.
+It's difficult to motivate people to use another language if they're using it only for one relatively small part of their job.
+Likewise, existing golang developers are not going to be drawn to contributing to Strimzi on account of a single golang component.
+So over all we expect our community would be better able to engage with the Canary component if it was implemented in Java.
+
+### Desirable features
+
+There are several client features we would like to be able use.
+
+* Support for SASL `OAUTHBEARER` mechanism. Sarama supports this.
+* Support for forcing leader elections. Sarama doesn't support this.
+* Support for transactional produce. Sarama doesn't seem to support this, at least not in a usable way. You can send the transaction-related requests, but there is no client library support akin to the `TransactionManager`, for ensuring the RPCs are sent correctly. 
+
+### Differential observations
+
+By running multiple producers and/or consumers, a canary could more accurately indicate where problems lay. 
+
+For example, in a cluster with multiple listeners using different authentication mechanisms, if produce requests via a `SASL/OAUTHBEARER`-authenticated listener fail, but succeed via the `SASL/PLAIN`-authenticated listener you can infer that the problem is with the login mechanism. 
+
+There are many dimensions in which you can do this: 
+
+* DNS resolution
+* TLS (e.g. different versions or detecting imminent broker certificate expiry
+* Producing to replicas on different volumes
+* With/without transactional producers, to flush out problems with transaction coordinators.
+* Similarly with/without using consumer groups. Using different group ids to detect problems with group coordinators.
+
+### Integration examples
+
+Currently the Canary provides a binary and nothing more. 
+This limits uptake of the canary because to use it for its intended purpose you need to configure prometheus scraping, alerting and dashboards to make sense of what the canary is telling you. 
+
+We would like to provide example of all these things, as we [do for the operator](https://github.com/strimzi/strimzi-kafka-operator/tree/main/examples/metrics).
+
+### Actionable documentation
+
+Assuming the differential observations feature proves to provide worthwhile information, we would proceed to also provide documented procedures recommending corrective actions based on the specific faults it reported.
```

Comment:
In the long term, some of these actions could be carried out automatically by Strimzi itself (assuming it is aware of a canary running).